srtcore/sync.h

```diff
@@ -625,19 +621,91 @@ class CGlobEvent
 public:
     /// Triggers the event and notifies waiting threads.
     /// Simply calls notify_one().
-    static void inline triggerEvent()
-    {
-        return g_Sync.notify_one();
-    }
+    static void triggerEvent();
 
     /// Waits for the event to be triggered with 10ms timeout.
     /// Simply calls wait_for().
-    static bool waitForEvent()
-    {
-        return g_Sync.lock_wait_for(milliseconds_from(10));
-    }
+    static bool waitForEvent();
 };
 
+////////////////////////////////////////////////////////////////////////////////
+//
+// CThread class
+//
+////////////////////////////////////////////////////////////////////////////////
+
+#ifdef USE_STDCXX_CHRONO
+typedef std::system_error CThreadException;
+using CThread = std::thread;
+#else // pthreads wrapper version
+typedef ::CUDTException CThreadException;
+
+class CThread
+{
+public:
+    CThread();
+    /// @throws std::system_error if the thread could not be started.
+    CThread(void *(*start_routine) (void *), void *arg);
+
+#if HAVE_FULL_CXX11
+    CThread& operator=(CThread &other) = delete;
+    CThread& operator=(CThread &&other);
+#else
+    CThread& operator=(CThread &other);
+#endif
+
+public: // Observers
+    /// Checks if the CThread object identifies an active thread of execution.
+    /// A default constructed thread is not joinable.
+    /// A thread that has finished executing code, but has not yet been joined
+    /// is still considered an active thread of execution and is therefore joinable.
+    bool joinable() const;
+
+public:
+    /// Blocks the current thread until the thread identified by *this finishes its execution.
+    /// If that thread has already terminated, then join() returns immediately.
+    ///
+    /// @throws std::system_error if an error occurs
+    void join();
+
+private:
+    pthread_t m_thread;
+};
+#endif
+
+/// This function should be used to do CThread assignments:
+/// @code
+/// CThread a();
+/// a = CThread(func, args);
+/// @endcode
+///
+/// @throws CThreadException if the thread could not be started.
+///
+#ifdef USE_STDCXX_CHRONO
+template< class Function >
+inline void SetThread(CThread& th, Function&& f, void* args)
+#else
+inline void SetThread(CThread& th, void* (*f) (void*), void* args)
+#endif
+{
+#if HAVE_FULL_CXX11 || defined(USE_STDCXX_CHRONO)
+    th = CThread(f, args);
+#else
+    CThread dummy(f, args);  // No move semantics in C++03
```

Comment:
Isn't it simpler and more reliable if you make a kinda "extension" to the API (that would be used exclusively by this function - might be also through friend/private) that will simply set the thread based on this function? I mean, simple redirection to a function avoids any nuisances of constructor/destructor, dealing with move semantix, temporary objects and alike.

Reply:
Added `create_thread(..)` function to do this.